Mathlib/LinearAlgebra/QuadraticForm/Basic.lean

```diff
@@ -500,397 +512,437 @@ end SemiringOperators
+variable [CommSemiring R] [AddCommMonoid M] [Module R M] [AddCommMonoid N] [Module R N]
+variable {M' : Type v} [AddCommMonoid M'] [Module R M']
 
-/-- Compose the quadratic form with a linear function. -/
-def comp (Q : QuadraticForm R N) (f : M →ₗ[R] N) : QuadraticForm R M where
+/-- Compose the quadratic map with a linear function on the right. -/
+def comp (Q : QuadraticMap R M' N) (f : M →ₗ[R] M') : QuadraticMap R M N where
   toFun x := Q (f x)
   toFun_smul a x := by simp only [map_smul, f.map_smul]
   exists_companion' :=
     let ⟨B, h⟩ := Q.exists_companion
     ⟨B.compl₁₂ f f, fun x y => by simp_rw [f.map_add]; exact h (f x) (f y)⟩
-#align quadratic_form.comp QuadraticForm.comp
+#align quadratic_form.comp QuadraticMap.comp
 
 @[simp]
-theorem comp_apply (Q : QuadraticForm R N) (f : M →ₗ[R] N) (x : M) : (Q.comp f) x = Q (f x) :=
+theorem comp_apply (Q : QuadraticMap R M' N) (f : M →ₗ[R] M') (x : M) : (Q.comp f) x = Q (f x) :=
   rfl
-#align quadratic_form.comp_apply QuadraticForm.comp_apply
+#align quadratic_form.comp_apply QuadraticMap.comp_apply
 
-/-- Compose a quadratic form with a linear function on the left. -/
+/-- Compose a quadratic map with a linear function on the left. -/
 @[simps (config := { simpRhs := true })]
-def _root_.LinearMap.compQuadraticForm [CommSemiring S] [Algebra S R] [Module S M]
-    [IsScalarTower S R M] (f : R →ₗ[S] S) (Q : QuadraticForm R M) : QuadraticForm S M where
+def _root_.LinearMap.compQuadraticMapAux (f : N →ₗ[R] M') (Q : QuadraticMap R M N) :
+    QuadraticMap R M M' where
   toFun x := f (Q x)
-  toFun_smul b x := by simp only [Q.map_smul_of_tower b x, f.map_smul, smul_eq_mul]
+  toFun_smul b x := by simp only [map_smul, f.map_smul]
   exists_companion' :=
     let ⟨B, h⟩ := Q.exists_companion
-    ⟨(B.restrictScalars₁₂ S S).compr₂ f, fun x y => by
-      simp_rw [h, f.map_add, LinearMap.compr₂_apply, LinearMap.restrictScalars₁₂_apply_apply]⟩
-#align linear_map.comp_quadratic_form LinearMap.compQuadraticForm
+    ⟨B.compr₂ f, fun x y => by simp only [h, map_add, LinearMap.compr₂_apply]⟩
+
+/-- Compose a quadratic map with a linear function on the left. -/
+@[simps! (config := { simpRhs := true })]
+def _root_.LinearMap.compQuadraticMap [CommSemiring S] [Algebra S R] [Module S N] [Module S M]
+    [IsScalarTower S R N] [IsScalarTower S R M] [Module S M']
+    (f : N →ₗ[S] M') (Q : QuadraticMap R M N) : QuadraticMap S M M' :=
+  _root_.LinearMap.compQuadraticMapAux f Q.restrictScalars
+#align linear_map.comp_quadratic_form LinearMap.compQuadraticMap
 
 end Comp
 
 section CommRing
 
-variable [CommSemiring R] [AddCommMonoid M] [Module R M]
+variable [CommSemiring R]  [NonUnitalNonAssocCommSemiring N]  [AddCommMonoid M] [Module R M]
+variable [Module R N] [SMulCommClass R N N] [IsScalarTower R N N]
 
 /-- The product of linear forms is a quadratic form. -/
-def linMulLin (f g : M →ₗ[R] R) : QuadraticForm R M where
+def linMulLin (f g : M →ₗ[R] N) : QuadraticMap R M N where
```

Comment:
I think `N` should probably be called `A` or something here, to indicate it is more than a module.

Reply:
Done, thanks.